tests/spec_decode/e2e/test_correctness.py

```diff
@@ -28,14 +28,116 @@
+    kwargs = {
+        **common_llm_kwargs,
+        **per_test_common_llm_kwargs,
+        **test_llm_kwargs,
+    }
+    output_len = 32
+    temperature = 0.0
+
+    llm = AsyncLLM(**kwargs)
+    prompts = [
+        "Hello, my name is",
+        "The president of the United States is",
+        "The capital of France is",
+        "The future of AI is",
+    ]
+
+    prompts = [prompt for prompt, _ in zip(cycle(prompts), range(batch_size))]
+
+    sampling_params = SamplingParams(
+        max_tokens=output_len,
+        ignore_eos=True,
+        temperature=temperature,
+    )
+
+    outputs = llm.generate(prompts, sampling_params)
+    batch_token_ids = [output.outputs[0].token_ids for output in outputs]
+    batch_tokens = [output.outputs[0].text for output in outputs]
+
+    # Expect a generation for each prompt in the batch.
+    assert len(batch_token_ids) == len(prompts)
+
+    # Expect each generation to have expected number of tokens (note ignore_eos
+    # is True).
+    assert [len(token_ids)
+            for token_ids in batch_token_ids] == ([output_len] * batch_size)
+
+    # Expect detokenized string to match.
+    tok = AutoTokenizer.from_pretrained("JackFram/llama-68m")
+    for actual_tokens, actual_token_ids in zip(batch_tokens, batch_token_ids):
+        expected_tokens = tok.decode(actual_token_ids)
+        print(f"{actual_token_ids=}")
+        assert actual_tokens.strip() == expected_tokens.strip()
+
+    del llm
+
+    # add tokens to result if success in subprocess
+    result["batch_tokens"] = batch_tokens
+
+
+@pytest.mark.parametrize(
+    "common_llm_kwargs",
+    [{
+        # Use a small model for a fast test.
+        # Note this is repeated in the test body; to initialize a tokenizer.
+        "model": "JackFram/llama-68m",
+
+        # Skip cuda graph recording for fast test.
+        "enforce_eager": True,
+
+        # Required for spec decode.
+        "use_v2_block_manager": True
+    }])
+@pytest.mark.parametrize("per_test_common_llm_kwargs", [
+    {
+        "speculative_model": "JackFram/llama-68m",
+        "num_speculative_tokens": 5,
+    },
+])
+@pytest.mark.parametrize("test_llm_kwargs", [{}])
+@pytest.mark.parametrize("batch_size", [1])
+@pytest.mark.parametrize("seed", [1])
+def test_spec_decode_e2e_with_detokenization_async(request, common_llm_kwargs,
+                                                   per_test_common_llm_kwargs,
+                                                   test_llm_kwargs,
+                                                   batch_size: int, seed: int):
+    """Run generation with speculative decoding on a batch. Verify the engine
+    generates the correct number of tokens (via ignore_eos=True), and that the
+    detokenization matches HF transformers.
+    """
+    import time
+    from multiprocessing import Manager, Process
+
+    with Manager() as manager:
+        result = manager.dict()
+        p = Process(target=_test_spec_decode_e2e_with_detokenization_async,
```

Comment:
This is prone to leak. Consider;

```
try:
    p = Process(target=_test_spec_decode_e2e_with_detokenization_async,
                    args=(request, common_llm_kwargs,
                          per_test_common_llm_kwargs, test_llm_kwargs,
                          batch_size, seed, result))
    p.start()
    p.join()
finally:
    p.terminate()
```

or I recommend you to use ray (and use ray.shutdown() in the finally). 

```
import ray
try:
    ray.init()
    @ray.remote
    def run():
        # call your thing
    ray.get(run.remote())
finally:
    ray.shutdown()
```